### PR TITLE
Provision refs resolver

### PR DESCRIPTION
**Pull request by longhotsummer**
Replace the old internal refs resolver with a more robust provision refs resolver that can also resolve provisions in external works.

The bulk of the tests were copied over from the old resolver, especially the edge cases. We now can resolve a lot more of these than before.

A single resolver supports multiple languages, currently English and Afrikaans. Adding new languages should hopefully be relatively simple.

https://www.loom.com/share/730117beb7c048e1b857c1914ee1952e

https://www.loom.com/share/12ac35fd26ab4c07aeb850252f6f1620

## Outstanding

* [x]  'subsection (1)(a) of section 27'

### CODE REVIEW

File: indigo/analysis/refs/provision_refs.peg [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                 # thereof
+  thereof     <- comma? WS* (thereof_en / thereof_af) %thereof
+
+  # --------
+  # terminals
+  # --------
+
+                 # english
+  unit_en     <- `articles` / `article` /
+                 `chapters` / `chapter` /
```
Comment (goose-life) [inline indigo/analysis/refs/provision_refs.peg:70]:
```suggestion
                 `chapters` / `chapter` /
                 `parts` / `part` /
```

Comment (goose-life) [reply indigo/analysis/refs/provision_refs.peg:70]: oops, that's not alphabetical

File: indigo/analysis/refs/provision_refs.peg [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                 `items` / `item` /
+                 `paragraphs` / `paragraph` /
+                 `points` / `point` /
+                 `sections` / `section` /
+                 `subparagraphs` / `subparagraph` /
+                 `subsections` / `subsection`
+
+                 # afrikaans
+  unit_af     <- `afdelings` / `afdeling` /
+                 `artikels` / `artikel`
```
Comment (goose-life) [inline indigo/analysis/refs/provision_refs.peg:80]:
```suggestion
                 `artikels` / `artikel`
                 `dele` / `deel`
```

File: indigo/analysis/refs/provision_refs.peg [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+  # --------
+  # terminals
+  # --------
+
+                 # english
+  unit_en     <- `articles` / `article` /
+                 `chapters` / `chapter` /
+                 `items` / `item` /
+                 `paragraphs` / `paragraph` /
```
Comment (goose-life) [inline indigo/analysis/refs/provision_refs.peg:72]:
```suggestion
                 `paragraphs` / `paragraph` /
                 `parts` / `part` /
```

File: indigo/analysis/refs/provision_refs.peg [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  _and        <- and_en / and_af
+  and_en      <- `and`
+  and_af      <- `en`
+
+  _or         <- or_en / or_af
+  or_en       <- `or`
+  or_af       <- `of`
+
+                 # note that "tot" must come before "to" because the latter is a substring of the former
+  to          <- to_af / to_en
```
Comment (goose-life) [inline indigo/analysis/refs/provision_refs.peg:96]: we might want to add `-` and `–` (and maybe even `—`) to this at some point

File: indigo/analysis/refs/provisions.py [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+    - name: Section
+    - ref: ProvisionRef("32", ..., ProvisionRef("(a)", ...))
+    - sub_refs: ProvisionRef("(b)", ...)
+    """
+    # eg "section" or "part"
+    name: str
+    # the primary ref
+    ref: ProvisionRef
+    # secondary refs that are resolved relative to the primary
+    sub_refs: List[ProvisionRef] = None
```
Comment (goose-life) [inline indigo/analysis/refs/provisions.py:50]: something complained about me using an object in a list like this the other day; is that because I didn't use `List` from `typing`?

File: indigo/analysis/refs/provisions.py [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+    element_names = {
+        # en
+        "article": "article",
+        "articles": "article",
+        "chapter": "chapter",
+        "chapters": "chapter",
+        "item": "item",
+        "items": "item",
+        "paragraph": "paragraph",
+        "paragraphs": "paragraph",
```
Comment (goose-life) [inline indigo/analysis/refs/provisions.py:65]:
```suggestion
        "paragraph": "paragraph",
        "paragraphs": "paragraph",
        "part": "part",
        "parts": "part",
```

File: indigo/analysis/refs/provisions.py [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+        "points": "point",
+        "section": "section",
+        "sections": "section",
+        "subparagraph": "subparagraph",
+        "subparagraphs": "subparagraph",
+        "subsection": "subsection",
+        "subsections": "subsection",
+        # af
+        "artikel": "article",
+        "artikels": "article",
```
Comment (goose-life) [inline indigo/analysis/refs/provisions.py:76]:
```suggestion
        "artikel": "article",
        "artikels": "article",
        "deel": "part",
        "dele": "part",
```

File: indigo/analysis/refs/provisions.py [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+        "hoofstuk": "chapter",
+        "hoofstukke": "chapter",
+        "paragraaf": "paragraph",
+        "paragrawe": "paragraph",
+        "punt": "point",
+        "punte": "point",
+        "afdeling": "section",
+        "afdelings": "section",
```
Comment (goose-life) [inline indigo/analysis/refs/provisions.py:84]: I'm sorry to be the one to tell you, but 'artikel' is used as the equivalent of 'section' 😅

File: indigo/analysis/refs/provisions.py (hunk 5 above)
Comment (longhotsummer) [reply indigo/analysis/refs/provisions.py:53]: > something complained about me using an object in a list like this the other day; is that because I didn't use `List` from `typing`?

Probably

File: indigo/analysis/refs/provisions.py (hunk 8 above)
Comment (longhotsummer) [reply indigo/analysis/refs/provisions.py:91]: BUT CHATGPT CAN DO NO WRONG EDITH!

**Comment by longhotsummer** [comment]
This now supports local internal refs! https://www.loom.com/share/d522674b73864b2f8f897d173acaccdf

File: indigo/analysis/refs/provisions.py [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+        "point": "point",
+        "points": "point",
+        "section": "section",
+        "sections": "section",
+        "subparagraph": "subparagraph",
+        "subparagraphs": "subparagraph",
+        "subsection": "subsection",
+        "subsections": "subsection",
+        # af
+        "artikel": "article",
```
Comment (longhotsummer) [inline indigo/analysis/refs/provisions.py:80]: `artikel` can be both `section` and `article`

**Comment by goose-life** [comment]
Very cool! 

Primary legislation that isn't deeply nested seems to work exactly as designed, including 'paragraph (b) of section 9(1) of blah Act num of year'. (I inserted that text in a document to test it, there might be ones like that in the wild but they weren't easy to find.)

/akn/bw/act/1999/18 (Agrochemicals Act) – no issues

A couple of edge cases:

#### /akn/za/act/1993/194 (Trade Marks Act, 1993)
- I think for these, the first part before 'of (sub)section' gets evaluated before the rest, and makes a false positive local match
  - in section 19(2), 'subsection (1)(a) of section 27' – only the (1) matches section 19(1)
    - I tried deleting all extra text, and checked section 27(1)(a) does exist
  - in section 30(2), 'subsection (1) of section 18' – the (1) matches section 30(1)
  - similarly in section 31(2)
  - in section 38(8)(b), 'paragraph (b) of subsection (6)' – the (b) matches the current paragraph
- 'paragraph' vs subsection mark-up (some missed)

#### /akn/bw/act/si/2020/61 (Emergency (COVID-19) Regulations, 2020)
- in section 5(3)(b)(ii), reference to 'paragraph (i)' matches section 7(2)(i)
  - probably 'subparagraph' would match — I've checked, and yes it does, so file this under 'non-matching element'
- in section 5(3)(b)(iv), reference to 'subparagraph (iii)(aa)' – only the (iii) matches section 5(3)(b)(i)(ee)(iii) 

#### /akn/za/act/1988/57 (Trust Property Control Act, 1988)
- in the table in the Schedule, sections of Acts being amended are linked (I think this is unavoidable, unless we just skip tables)

#### /akn/za/act/1949/40 (Transfer Duty Act, 1949)
- in the definition of "property", reference to 'paragraph (b) or (c)' matches section 4(3)(b) and section 5(7)(c)
- 'sub-section', 'section one' passed over

#### non-matching element names
Where we've used subsection, or subparagraph, and they've used 'paragraph', internal references are missed.
This can't always be solved by using the elements they do in the text, e.g. AKN doesn't support regulations or subregulations, so we have to use sections and subsections for those.
Sub-section is also missed.

#### Cap. references
I had a note that these weren't picked up, but I can think of a couple of other reasons why that wouldn't have worked, so I'd have to test it more carefully.
No false positives here, so not blocking.

#### Other missed references
- In Botswana, for example, they don't use Act or Cap. numbers — just short titles. It would be cool to eventually be able to find those too.
- Some traditions use 'section one' rather than 'section 1'.

**Comment by longhotsummer** [comment]
> in section 5(3)(b)(iv), reference to 'subparagraph (iii)(aa)' – only the (iii) matches section 5(3)(b)(i)(ee)(iii)

This should now be fixed.

**Comment by longhotsummer** [comment]
> probably 'subparagraph' would match — I've checked, and yes it does, so file this under 'non-matching element'

I think we could safely expand the search space for some of these -- eg. 'paragraph' should search for 'paragraph' and 'subparagraph' (and maybe subsection too).

**Comment by longhotsummer** [comment]
> 'paragraph' vs subsection mark-up (some missed)

Do you think it's safe to make something like 'paragraph (16)' look for `paragraph`, `subsection` and `subparagraph` elements? And probably equivalently for `subsection` and `subparagraph`.

**Comment by longhotsummer** [comment]
> 'sub-section', 'section one' passed over

I fixed `sub-section` but not `section one`.

**Comment by goose-life** [comment]
> > 'paragraph' vs subsection mark-up (some missed)
> 
> Do you think it's safe to make something like 'paragraph (16)' look for `paragraph`, `subsection` and `subparagraph` elements? And probably equivalently for `subsection` and `subparagraph`.

I want to say yes? Especially since you seem to have solved the other issues around where to look first, and finding the 'of section 2(3)' matches before their preceding 'paragraph (a)'.

I suppose it does open us up to potentially more false matches, but if it's a question of just adding more elements to look for I think we can try it out and see? (rather than a big rewrite)

Here are my suggestions from my test drive:
- 'paragraph' – look for `paragraph`, `subparagraph`, `subsection` (in that order of preference, if that's a thing)
- 'subparagraph' – look for `subparagraph`, `paragraph`
- 'regulation' – look for `section`
- 'subregulation' – look for `subsection`, `paragraph`, `subparagraph`

File: indigo/tests/test_provision_refs.py [hunk 10]
```diff
@@ -0,0 +1,1248 @@
...
+                </paragraph>
+                <paragraph eId="sec_1__subsec_3__para_b">
+                  <num>(b)</num>
+                </paragraph>
+              </subsection>
+              <subsection eId="sec_1__subsec_4">
+                <num>(4)</num>
+                <content>
+                  <p>referred to in sub-section <ref href="#sec_1__subsec_1">(1)</ref>, <ref href="#sec_1__subsec_2">(2)</ref> or <ref href="#sec_1__subsec_3__para_b">(3)(b)</ref>, to the extent that...</p>
+                  <p>For the purposes of subsection <ref href="#sec_2__subsec_3__para_b">(1)(b)</ref> of section <ref href="sec_2">2</ref> the intention is that</p>
```
Comment (goose-life) [inline indigo/tests/test_provision_refs.py:551]:
```suggestion
                  <p>For the purposes of subsection <ref href="#sec_2__subsec_1__para_b">(1)(b)</ref> of section <ref href="sec_2">2</ref> the intention is that</p>
```

**Comment by longhotsummer** [comment]
Support for synonyms, regulations and sub-regulations added.

**Comment by longhotsummer** [comment]
image

**Comment by longhotsummer** [comment]
The bulk of the raised issues are fixed. I had to disable the Afrikaans translation of 'or'. I don't have a holistic fix currently for indicating to the parser which language it should be working in.

File: indigo/analysis/refs/provisions.py [hunk 11]
```diff
@@ -0,0 +1,555 @@
...
+        sections 18, 61 and 62(1).
+        in terms of section 2 or 7
+        A person who contravenes sections 4(1) and (2), 6(3), 10(1) and (2), 11(1), 12(1), 19(1), 19(3), 20(1), 20(2), 21(1), 22(1), 24(1), 25(3), (4) , (5) and (6) , 26(1), (2), (3) and (5), 28(1), (2) and (3) is guilty of an offence.
+    """
+    xml_marker_tag = "ref"
+    xml_ancestor_xpath = '|'.join(f'//ns:{x}'
+                                  for x in ['coverpage', 'preface', 'preamble', 'body', 'mainBody', 'judgmentBody', 'conclusions'])
+    xml_candidate_xpath = (".//text()[not("
+                           "ancestor::ns:ref or ancestor::ns:heading or ancestor::ns:subheading or ancestor::ns:num "
+                           "or ancestor::ns:embeddedStructure or ancestor::ns:quotedStructure)]")
```
Comment (goose-life) [inline indigo/analysis/refs/provisions.py:290]:
```suggestion
                           "or ancestor::ns:embeddedStructure or ancestor::ns:quotedStructure or ancestor::ns:table)]")
```

Comment (goose-life) [reply indigo/analysis/refs/provisions.py:290]: a bit rough n ready, but will avoid us marking up section references in tables where the amended Act is given in a different cell (normally in the first column)

File: indigo/analysis/refs/provisions.py [hunk 12]
```diff
@@ -0,0 +1,555 @@
...
+    element_names = {
+        # en
+        "article": "article",
+        "articles": "article",
+        "chapter": "chapter",
+        "chapters": "chapter",
+        "item": "item",
+        "items": "item",
+        "paragraph": ["paragraph", "subparagraph", "subsection"],
+        "paragraphs": ["paragraph", "subparagraph", "subsection"],
```
Comment (goose-life) [inline indigo/analysis/refs/provisions.py:152]:
```suggestion
        "paragraph": ["paragraph", "subparagraph", "item", "subsection"],
        "paragraphs": ["paragraph", "subparagraph", "item", "subsection"],
```

Comment (goose-life) [reply indigo/analysis/refs/provisions.py:152]: for paragraphs of defined terms

**Review by goose-life — APPROVED** [review]
two minor suggestions, happy for you to leave them out (especially tables) if you think it best

File: indigo/analysis/refs/provisions.py (hunk 11 above)
Comment (longhotsummer) [reply indigo/analysis/refs/provisions.py:290]: added

**Comment by longhotsummer** [comment]
@goose-life I've added `crossHeading` to the list of elements to ignore, too.
